### PR TITLE
Add an option for a clean debug build

### PR DESCRIPTION
**Pull request by noctux**
The Makefile now separates objects of regular and debug builds into the
two subfolders `./build` and `./build-debug` and links those two two
different binaries `./{pod,news}boat` and `./{pod,news}boat-dbg`.

Builds for the debug binaries can be invoked either by `make debug` or
by setting the `DEBUG=1` variable in any regular make invocation, e.g.
`make DEBUG=1 newsboat`.

What is currently missing is a good idea on how to handle the `clean` targets.
`make clean` works for the regular binaries/objects and a regular 
`make DEBUG=1 clean{,-.*}` works for the non debug variants, but that is
hardly a proper solution in my opinion.

### CODE REVIEW

**Comment by Minoru** [comment]
Thanks! If I understand correctly, this PR tackles two problems:
1. how to build debug binaries (no optimizations, debug info included);
2. how to make sure that debug object files aren't mixed with the release ones.

Is that right?

The first problem is solved by appending `-O0 -ggdb` to `CXXFLAGS`, which looks good to me. If you extract that feature into a separate commit, I'll cherry-pick it. Or just commit it to master yourself, it's small and simple enough.

The second problem is solved by introducing something akin to out-of-tree builds. I wonder if this can be done simpler; for example, Makefile might compare `CXXFLAGS` to a file written earlier, rewrite the file if flags changed, and rebuild *everything* when the file changes. This way, we keep in-tree builds while guaranteeing that debug objects won't be mixed with non-debug ones. Does this make sense?

Overall, though, I am not sure this is a good investment of anyone's effort. More high-level build systems (e.g. CMake) already implement that sort of thing, we only need to move. Perhaps spending some time on #14 is a better idea?

**Comment by noctux** [comment]
Hey,

On Sun, Mar 04, 2018 at 08:31:23PM +0000, Alexander Batischev wrote:
> Thanks! If I understand correctly, this PR tackles two problems:
>
> 1. how to build debug binaries (no optimizations, debug info included);
> 2. how to make sure that debug object files aren't mixed with the release ones.
> 
> Is that right?

Yes, this is correct.

> The first problem is solved by appending `-O0 -ggdb` to `CXXFLAGS`,
> which looks good to me. If you extract that feature into a separate
> commit, I'll cherry-pick it. Or just commit it to master yourself,
> it's small and simple enough.

I've split the commit now, however, imho this change, in isolation, adds
more problems than it solves, as people (or at least me, tbh) are bound
to forget the mandatory `make clean`.

> The second problem is solved by introducing something akin to
> out-of-tree builds. I wonder if this can be done simpler; for example,
> Makefile might compare `CXXFLAGS` to a file written earlier, rewrite
> the file if flags changed, and rebuild *everything* when the file
> changes. This way, we keep in-tree builds while guaranteeing that
> debug objects won't be mixed with non-debug ones. Does this make
> sense?

It sort of makes sense, however I've got two objections: This workflow
of comparing the CXXFLAGS-file feels like explicitly working against the
Make-way of doing things. Suddenly, we leave the neat, timestamped world
of %.o:%.cpp. However, this is mostly a style thing (even though it
might be tricky to select the target it should apply to, etc.pp.).

What I consider worse is the debugging experience: Consider an issue
that forces you to jump between release and debug-binaries while
investigating: Each small change forces you to do two full
recompilations of all objects, instead of just recompiling the affected
objects and relinking.

> Overall, though, I am not sure this is a good investment of anyone's
> effort. More high-level build systems (e.g. CMake) already implement
> that sort of thing, we only need to move. Perhaps spending some time
> on #14 is a better idea?

This is quite subjective, but I personally do not consider a 350loc
Makefile to be a real problem, especially since it seems to be quite
straightforward to read (even though config.sh is a bit of a hack).
CMake and other higher-level buildsystems promise to abstract away this
library-resolution in a platform-independent manner. However, this layer
of indirection in my experience makes debugging a lot harder, because if
there is a library resolution problem, one is not looking at ~5loc of
bash but on an monstrosity of CMake-macros, so I'd personally take the
more debugable - though maybe a bit uglier - Makefile any day. (Of
course, there is some limit to the complexity a Makefile should possess,
so I would never argue that Kbuild-like beasts were a good idea...)

So long story short: I personally would prefer the Makefile solution, so
if you've got some concrete issues, I'd be happy to try to squeeze those
into the current Makefile.

**Comment by Minoru** [comment]
> I've split the commit now, however, imho this change, in isolation, adds more problems than it solves, as people (or at least me, tbh) are bound to forget the mandatory `make clean`.

Don't you run into the same problem when you build with `PROFILE=1`? I personally develop with profiling all the time (otherwise tests won't run), and release build is such a big event for me that I never forget to clean the tree. Not so for you?

> This workflow of comparing the CXXFLAGS-file feels like explicitly working against the Make-way of doing things

Agreed. I won't cling to that idea then, it's really not that good :)

> Consider an issue that forces you to jump between release and debug-binaries while investigating

A build system that supports out-of-tree builds will let you have as many different builds as you want, each configured the way you want.

I never actually ran into the issue you describe, but if I only had Makefiles and needed two separate builds, I'd take a clean tree, `cp -rvl` it somewhere, and do two separate builds that way. The sources will stay in sync because of hardlinks, but I'll be able to configure each tree differently.

> However, this layer of indirection in my experience makes debugging a lot harder, because if there is a library resolution problem, one is not looking at ~5loc of bash but on an monstrosity of CMake-macros

I agree, but consider the other side: CMake is used by a lot of people, which means that:

* for any given problem, there's bigger potential pool of people who might help, including people who stumbled into exact same issue and posted about it somewhere. The only people who know anything about our custom Makefile are us, and we aren't many;

* CMake is ran more often than our Makefile will be, ever, so we're less likely to run into any problems at all—they would've been fixed by now.

I don't push for CMake specifically (though it's slowly becoming a standard, it seems), but I do think we need to replace our Makefile with something more high-level that both configures and builds. Makefile only covers the latter part.

> I personally would prefer the Makefile solution, so if you've got some concrete issues, I'd be happy to try to squeeze those into the current Makefile.

The problem I have is with squeezing stuff *out* of the Makefile :) I still need to look into package managers as discussed in #89—that would rid us of config.sh, and should make it easier to keep on top of Catch2 and json.hpp releases. As for the build systems, I'm waiting for newer CMake releases to hit the distros to re-try #14, and keeping an eye out for more systems to try the next time I want to play with this. If you're interested in any of that, you can try some and report on your experience in one of the issues (or a new one; you know how this "issue tracking" thingy works :)

**Comment by Minoru** [comment]
I'm closing this PR because I do not intend to merge these changes—sorry about that, @noctux. Hope my comments conveyed the reasons clearly enough. I'm still open to discussion about build systems here and elsewhere.
